### DISCUSSION TITLE
Export of notes and child-notes

### DISCUSSION THREAD

**Discussion by zottelsheep** [zk-org/zk#232] [Ideas]
I'm currently using zk as a way to organize my university-notes and linked topics kinda like a knowledge-base using wiki-links with pipe-syntax `[[link|title]]`.
What I'd like to do, is to export a single note as a pdf (in my case pandoc -> latex -> pdf) or parseable markdown-document. Pandoc has currently no support for Wiki-Links although there seems to be some effort regarding [this](https://github.com/jgm/pandoc/pull/7705). I'm not sure if this would parse the wiki-link to the title of the linked note, if none is explicitly given. And zk could definitely do that (The parsing stuff).

I have searched the documentation of zk but found nothing except for [process-the-content-of-a-note](https://github.com/mickael-menu/zk/blob/main/docs/external-processing.md#process-the-content-of-a-note) but that just throws the raw-content into stout if I'm not mistaken.

Some ideas I had concerning that topic.
- Export single note and parse Wiki-links to title name. Either from title of the linked note itself or an explicitly given one.
- Export child-notes of a given depth and link them within the exported file either using
    - Concatenation and internal links
    - Export as separate files inside same directory and link them using relative links
    - Maybe general approach: Function that takes a list of notebook-files 
         - If they reference each other, create some kind of direct link between them (depending on concatenation or separate file approach). 
         - Solve to title for rest of linking
- resolve relative paths of normal file-links to either
    - export-location
    - notebook-root
- Absolute-Path Mode: Convert all relative paths inside a note to absolute paths

What are some thoughts on this. Is this something that would be interesting? I for one, think the concat-feature would be really cool and useful, since it allows for easy sharing of a note-stack or learning using a (more or less) coherent single document.

**Comment by mickael-menu** [comment]
Hi,

> * Export single note and parse Wiki-links to title name.

I'm not sure what you mean by that, do you have a concrete example?

> * Concatenation and internal links

This is something I've been thinking about, to create a draft document from an outline of links. But this is not high on my priority list right now.

**Reply by zottelsheep** [reply]
For Example: You have three Documents called 
- `1111-note_taking` with title "note taking"
- `2222-personal_knowledge_mangement` with title "Personal knowledge management"
- `3333-zettelkasten` with title "Zettlekasten"

The document `3333-zettelkasten` contains following text:
```markdown
# Zettelkasten
The zettelkasten is a system of [[1111]] and [[2222|personal knowledge management]] used in research and study.
```
Now i want to export `3333-zettelkasten` as a standalone note. ZK now parses its wikilink-syntax and inserts the title either from the note in the database or the one explicitly set withing the document:
``` markdown
# Zettelkasten
The zettelkasten is a system of _note taking_ and _personal knowledge management_ used in research and study.
```

I hope this clarifies it.

**Reply by mickael-menu** [reply]
Thanks, I understand better. I think an export command would be very useful, not only for sharing but also to backup portions of a notebook.

Instead of providing specific note(s) as arguments, we can use the same [filtering options](https://github.com/mickael-menu/zk/blob/main/docs/note-filtering.md) as `zk list` and `zk edit` to select the notes to export. This can be used to:
* Export one or several explicit notes or folders.
* Export one note [along the notes linked by it (or linking to it)](https://github.com/mickael-menu/zk/blob/main/docs/note-filtering.md#explore-links). This one will need some adjusting to also get the root note, but it could look like this:
    * `zk export --linked-by note.md --recursive --max-distance 2`
* Use `fzf` (with `--interactive`) to choose interactively the notes to export.
* Reuse the [shared note filters](https://github.com/mickael-menu/zk/blob/main/docs/config-filter.md).
* Export the notes with a given tags.
* etc.
 
I think we can split this feature in several steps/levels of complexity that can be tackled independently:

* **Level 0**: A simple copy of the selected notes which follows the folder structure of the notebook (to avoid breaking links). Links to notes that are not included are replaced by the title like in your example.
* **Level 1**: New `--format` flag to choose the export format, among:
    * `default/none` – keeps the notes as is.
    * `markdown` – replaces special syntax to regular markdown, for example wiki links to standard Markdown links.
    * other formats could be added later on: `html`, `pdf`, `latex`
        * `zk` could rely on Pandoc under the hood for the conversions.
* **Level 2**: New `--merge` or `--concat` flag that will merge all the notes into a single Markdown (or HTML, PDF, etc.) document.
    * We'll need to figure out the best strategy for ordering the notes in the document, and/or maybe expose several strategies including custom order.
    * Internal links will need to be transformed to anchors in the document.
* **Level 3**: An outlining mode, based on the `--merge` flag, to quickly draft a paper. For example you create a new note:
    ```md
    # An interesting subject to talk about

    ## [[i3hr|Main idea]]

    ### [[9tnw|Point 1]]

    ### [[5exr|Point 2]]

    ## [[i2kb]]
    ```
    `zk` would output a single note, merging all the linked notes and adjusting the Markdown heading levels to the ones in the outline.

> Absolute-Path Mode: Convert all relative paths inside a note to absolute paths

I don't really see the use case for absolute paths, do you have an example?

> If you are not against such functionality, I would love to help implement this. I wanted to try out Go for a while now. I am a bit swamped with university and work atm. so I probably wouldn't be able to start until end of August.

Sure! I'm always open to contributions as long as we agree on how it will look like together first. Let me know if and when you want to tackle this, I can give you pointers through the code.

**Comment by zottelsheep** [comment]
If you are not against such functionality, I would love to help implement this. I wanted to try out Go for a while now. I am a bit swamped with university and work atm. so I probably wouldn't be able to start until end of August.

**Comment by skervim** [comment]
Are there any updates on this idea? 

I think it would be fantastic to have an export command!

Especially it comes to my mind that this could help implement the principle of [separation of content and presentation](https://en.wikipedia.org/wiki/Separation_of_content_and_presentation): the notebook contains the content in an agnostic fashion  and different export choices can create diverse presentation views of this content.

**Reply by mickael-menu** [reply]
I don't have enough time to work on `zk` lately, but feel free to chime in on the suggestions: https://github.com/mickael-menu/zk/discussions/232#discussioncomment-3165614
